File: 2024/tiktok/1106.py
def getMaxRequest(bandwidth, request, total_bandwidth):
    n = len(bandwidth)
    dp = [0] * (total_bandwidth + 1)
    for i in range(n):  # select i-th request
        # calc bandwidth that serves max request
        for b in range(total_bandwidth, bandwidth[i] - 1, -1):
            dp[b] = max(dp[b], dp[b - bandwidth[i]] + request[i])
    return dp[total_bandwidth]


assert getMaxRequest(
    [200, 100, 350, 50, 100],
    [270, 142, 450, 124, 189],
    500
) == 763

assert getMaxRequest(
    [100, 500, 80, 25, 400],
    [100, 1000, 120, 110, 100],
    200
) == 230


def getMessageStatus(timestamps, messages, k):
    history = {}
    res = []
    for ts, msg in zip(timestamps, messages):
        if msg not in history:
            history[msg] = [ts]
            res.append("true")
        else:
            if history[msg][-1] + k > ts:
                res.append("false")
            else:
                res.append("true")
            history[msg].append(ts)
    return res


assert getMessageStatus(
    [1, 1, 1, 11],
    ["messsage-2", "messsage-2", "messsage-3", "messsage-2"],
    5
) == ["true", "false", "true", "true"]


MOD = 10**9+7
def maximumEfficiency(memory):
    n = len(memory)
    for i in range(n//2):
        ni = n - i
        m1, m2 = memory[i], memory[ni-1]
        memory[i], memory[ni-1] = min(m1, m2), max(m1, m2)
    return sum([memory[i]*(i+1) for i in range(n)]) % MOD

assert maximumEfficiency(
    [2,4,1,3]
) == 28

assert maximumEfficiency(
    [5,4,1,5,3,2]
) == 81

File: 2024/tiktok/1106.ts
function getMaxRequests(bandwidth: number[], request: number[], total_bandwidth: number): number {
    // Write your code here
    const n = bandwidth.length
    const dp = Array(total_bandwidth+1).fill(0)
    for (let i = 0; i < n; i++) {
        for (let b = total_bandwidth; b >= bandwidth[i]; b--) {
            dp[b] = Math.max(dp[b], dp[b-bandwidth[i]] + request[i])
        }
    }
    return dp[total_bandwidth]
}


function getMessageStatus(timestamps: number[], messages: string[], k: number): string[] {
    // Write your code here
    const history: Record<string, Array<number>> = {}
    const res: string[] = []
    for (let i = 0; i < timestamps.length; i++) {
        const ts = timestamps[i]
        const msg = messages[i]

        if (! (msg in history)) {
            history[msg] = [ts]
            res.push('true')
        } else {
            let msglen = history[msg].length
            if (history[msg][msglen-1] + k > ts) {
                res.push('false')
            } else {
                res.push('true')
            }
            history[msg].push(ts)
        }
    }
    return res
}

const MOD = 10**9+7
function maximumProductSum(memory: number[]): number {
    // Write your code here
    const n = memory.length
    for (let i = 0; i < Math.floor(n / 2); i++) {
        const ni = n-i
        const m1 = memory[i]
        const m2 = memory[ni-1]
        memory[i] = Math.min(m1, m2)
        memory[ni-1] = Math.max(m1, m2)
    }

    let sum = 0
    for (let i = 0; i < n; i++) {
        sum += memory[i] * (i+1)
        sum %= MOD
    }
    return sum
}
